printf/printf.1
.\"
.\" Copyright (c) 2005 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.\" Sccsid @(#)printf.1	1.2 (gritter) 7/17/05
.TH PRINTF 1 "7/17/05" "Heirloom Toolchest" "User Commands"
.SH NAME
printf \- print a text string
.SH SYNOPSIS
\fBprintf\fR \fIformat\fR [\fIargument\fR ...]
.SH DESCRIPTION
The
.I printf
command writes a string to standard output
with a controlled
.IR format .
It is essentially a utility variant of the `C' language function
.IR printf (3).
.PP
Most characters in the
.I format
string
are simply copied.
The only exceptions are the backslash `\e' and percent `%' characters.
.PP
The backslash `\e' character in the
.I format
string
serves as an escape for the following character.
The sequences
`\ea' (alert),
`\eb' (backspace),
`\ef' (formfeed),
`\en' (newline),
`\er' (carriage return),
`\et' (tabulator),
and `\ev' (vertical tabulator)
cause the corresponding control characters to be printed.
The sequence `\e\fInnn\fR',
where
.I nnn
is a one-, two-, or three-digit octal number,
cause the character (byte) with the corresponding number
to be printed.
`\e\e' prints the backslash character itself.
.PP
The percent `%' character in the
.I format
string introduces a conversion specification
consisting of up to five parts,
of which the first four are optional:
.TP 14
.I position
A positive decimal integer followed by a dollar `$' character
that specifies the
.I argument
to be used instead of the next unused one
for the following conversions.
.TP 14
.I flags
Zero or more of
.RS
.PD 0
.TP 8
.B \-
left-justify the field
.TP 8
.B +
always include a sign when printing a number
.TP
<space>
prefix the result by a space if there is no sign
.TP 8
.B #
alternate format
.TP 8
.B 0
pad numbers with leading zeros
.PD
.RE
.TP 14
.I field width
A decimal number giving the minimum width of the output printed.
The output is padded if necessary,
as controlled by the flags described above.
By default, no padding is performed.
If the field width is `*',
it is taken from the next unused
.IR argument .
.TP 14
.I precision
A dot `.',
followed by a decimal number giving
the minimum digits written for integer numeric conversions,
the minumum digits after the radix character
for floating-point numeric conversions,
or maximum bytes written for string conversions.
If the precisision is `.*',
it is taken from the next unused
.IR argument .
.TP 14
.I specifier
A character controlling the type of the conversion performed:
.RS
.PD 0
.TP 5
.B s
The next unused
.I argument
is written as a plain string.
.TP 5
.B b
The next unused
.I argument
is written as a string containing escape sequences
similar to those of
.IR echo (1).
The backslash sequences described above are supported,
except that the octal number must be prefixed by a zero
as in `\e0\fInnn\fR'.
A `\ec' sequence causes
.I printf
to exit immediately when it is encountered.
.TP 5
.B c
The first character (byte) of the next unused
.I argument
is written.
Any following characters in this argument are ignored.
.TP 5
.BR d ", i
The next unused
.I argument
is written as a signed decimal number.
.TP 5
.B o
The next unused
.I argument
is written as an octal number.
With the `#' flag,
it is prefixed by `0'.
.TP 5
.B u
The next unused
.I argument
is written as an unsigned decimal number.
.TP 5
.B x
The next unused
.I argument
is written as a hexadecimal number,
using lowercase characters.
With the `#' flag,
it is prefixed by `0x'.
.TP 5
.B X
The next unused
.I argument
is written as a hexadecimal number,
using uppercase characters.
With the `#' flag,
it is prefixed by `0X'.
.TP 5
.B f
The next unused
.I argument
is written as a floating-point number
in the style `[\-]\fInnn.nnn\fR'.
.TP 5
.B e
The next unused
.I argument
is written as a floating-point number
in the style `[\-]\fIn.nnn\fRe[+|\-]\fInn'.
.TP 5
.B E
The next unused
.I argument
is written as a floating-point number
in the style `[\-]\fIn.nnn\fRE[+|\-]\fInn'.
.TP 5
.B g
The next unused
.I argument
is written as a floating-point number
like an integer if there is no fractional part,
as described for `f' if the exponent is small,
or as described for `g' if the exponent is large.
.TP 5
.B G
The next unused
.I argument
is written as a floating-point number
like an integer if there is no fractional part,
as described for `f' if the exponent is small,
or as described for `G' if the exponent is large.
.TP 5
.B %
A percent character is printed.
No
.I argument
is consumed.
.PD
.RE
.PP
If the argument for the numeric specifiers starts with a
single- or double quote (`'\fIc\fR' or `"\fIc\fR'),
the numeric value of the following character (byte sequence)
in the current character encoding is used.
.PP
If the
.I format
string consumes at least an
.IR argument ,
no format specification contains a
.I position
part,
but there are still unused
.I arguments
after the entire format string has been evaluated once,
it is evaluated repeatedly until all arguments are consumed.
Missing
.I arguments
default to the empty string for string conversions,
and to zero if a numeric value is expected.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_CTYPE
Determines the mapping of bytes to characters
for `'\fIc\fR' and `"\fIc\fR'.
.SH "SEE ALSO"
echo(1),
printf(3)


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

printf/printf.c
/*
 * printf - print a text string
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, June 2005.
 */
/*
 * Copyright (c) 2005 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)printf.c	1.7 (gritter) 7/17/05";

#include <stdio.h>
#include <stdlib.h>
#include <libgen.h>
#include <locale.h>
#include <wchar.h>
#include <limits.h>
#include <errno.h>
#include "asciitype.h"

#if defined (__GLIBC__) && defined (_IO_getc_unlocked)
#undef	putchar
#define	putchar(c)	_IO_putc_unlocked(c, stdout)
#endif

static char		*fp;		/* format pointer */
static int		a;		/* current argument index */
static int		ab;		/* beginning of arguments */
static int		ac;		/* argc to main() */
static char		**av;		/* argv to main() */
static int		c;		/* last character (byte) read */
static const char	*progname;	/* argv[0] to main() */
static int		status;		/* exit status */
static int		mb_cur_max;	/* MB_CUR_MAX */
static int		dolflag;	/* n$ field encountered */

static void
usage(void)
{
	fprintf(stderr, "Usage: %s format [[[arg1] arg2] ... argn]\n",
			progname);
	exit(2);
}

#define	getnum(T, type, func)	static type \
T(const char *cp) \
{ \
	char	*xp; \
	wchar_t	wc; \
	int	i; \
	type	n; \
\
	if (*cp == '"' || *cp == '\'') { \
		if (mb_cur_max > 1 && cp[1] & 0200) { \
			if ((i = mbtowc(&wc, &cp[1], mb_cur_max)) < 0) \
				return WEOF; \
			return wc; \
		} else \
			return cp[1] & 0377; \
	} \
	errno = 0; \
	n = func(cp, &xp); \
	if (errno) { \
		fprintf(stderr, "%s: \"%s\" arithmetic overflow\n", \
			progname, cp); \
		status |= 1; \
		xp = ""; \
	} \
	if (*xp) { \
		fprintf(stderr, "%s: \"%s\" %s\n", progname, cp, \
			xp > cp ? "not completely converted" : \
				"expected numeric value"); \
		status |= 1; \
	} \
	return n; \
}

#define	getint(a, b)	strtol(a, b, 0)
#define	getuns(a, b)	strtoul(a, b, 0)
#define	getdouble(a, b)	strtod(a, b)

getnum(integer, int, getint)
getnum(unsgned, unsigned, getuns)
getnum(floating, double, getdouble)

static int
backslash(int bflag, int really)
{
	int	c, i, n, z = 1;

	fp++;
	if (mb_cur_max > 1 && *fp & 0200) {
		if ((n = mblen(fp, mb_cur_max)) < 0)
			n = 1;
	} else
		n = 1;
	switch (*fp) {
	case '\0':
		n = 0;
		/*FALLTHRU*/
	case '\\':
		if (really) putchar('\\');
		break;
	case 'a':
		if (really) putchar('\a');
		break;
	case 'b':
		if (really) putchar('\b');
		break;
	case 'c':
		if (bflag) {
			if (really)
				exit(status);
			else {
				while (*fp)
					fp++;
				return 0;
			}
		}
		goto dfl;
	case 'f':
		if (really) putchar('\f');
		break;
	case 'n':
		if (really) putchar('\n');
		break;
	case 'r':
		if (really) putchar('\r');
		break;
	case 't':
		if (really) putchar('\t');
		break;
	case 'v':
		if (really) putchar('\v');
		break;
	case '0':
		if (bflag) {
			if (fp[1]) {
				fp++;
				goto digit;
			}
			if (really) putchar('\0');
			break;
		}
		/*FALLTHRU*/
	case '1': case '2': case '3':
	case '4': case '5': case '6': case '7':
		if (bflag)
			goto dfl;
	digit:	c = 0;
		for (i = 0; i < 3 && octalchar(fp[i] & 0377); i++)
			c = c << 3 | (fp[i] - '0');
		if (really) putchar(c);
		n = i;
		break;
	default:
	dfl:	for (i = 0; i < n; i++)
			if (really) putchar(fp[i] & 0377);
		z = n;
	}
	fp += n - 1;
	return z;
}

static void
bconv(int width, int prec, char *sp)
{
	char	*ofp = fp;
	int	i, n, really = 1;

	fp = sp;
	if (width > 0) {
		really = 0;
		goto try;
	prt:	really = 1;
		fp = sp;
		for (i = 0; i < width - n && i + n < prec; i++)
			putchar(' ');
	}
try:	for (n = 0; *fp && n < prec; fp++) {
		switch (*fp) {
		case '\\':
			n += backslash(1, really);
			break;
		default:
			if (really) putchar(*fp & 0377);
			n++;
		}
	}
	if (width > 0 && really == 0)
		goto prt;
	fp = ofp;
	if (width < 0) {
		while (n < prec && n++ < -width)
			putchar(' ');
	}
}

#define	nextarg()	(a < ac ? av[a++] : "")
static void
percent(void)
{
	char	*fmt = fp, *sp;
	int	width = 0, prec = LONG_MAX;
	int	n;
	double	f;
	int	c;
	int	star = 0;
	int	sign = 1;

	if (*++fp == '\0') {
		fp--;
		return;
	}
	if (digitchar(*fp&0377)) {
		n = 0;
		for (sp = fp; digitchar(*sp&0377); sp++)
			n = n * 10 + *sp - '0';
		if (*sp == '$') {
			a = n + ab - 1;
			dolflag = 1;
			fmt = sp;
			fmt[0] = '%';
			fp = &sp[1];
		}
	}
loop:	switch (*fp) {
	case '-':
		sign = -1;
		/*FALLTHRU*/
	case '+':
	case '#':
	case '0':
	case ' ':
		fp++;
		goto loop;
	}
	if (digitchar(*fp&0377)) {
		do
			width = width * 10 + *fp++ - '0';
		while (digitchar(*fp&0377));
	} else if (*fp == '*') {
		width = a < ac ? integer(av[a++]) : 0;
		fp++;
		star |= 1;
	}
	width *= sign;
	if (*fp == '.') {
		fp++;
		if (digitchar(*fp&0377)) {
			prec = 0;
			do
				prec = prec * 10 + *fp++ - '0';
			while (digitchar(*fp&0377));
		} else if (*fp == '*') {
			prec = a < ac ? integer(av[a++]) : 0;
			fp++;
			star |= 2;
		}
	}
	switch (*fp) {
	case 'b':
		bconv(width, prec, nextarg());
		return;
	case '%':
		putchar('%');
		return;
	case 'd':
	case 'i':
	case 'o':
	case 'u':
	case 'x':
	case 'X':
	case 'c':
		n = *fp == 'c' ? *(nextarg()) & 0377 :
			*fp == 'd' || *fp == 'i' ?
				integer(nextarg()) :
				unsgned(nextarg());
		c = fp[1];
		fp[1] = '\0';
		switch (star) {
		case 3:
			printf(fmt, width, prec, n);
			break;
		case 2:
			printf(fmt, prec, n);
			break;
		case 1:
			printf(fmt, width, n);
			break;
		default:
			printf(fmt, n);
		}
		fp[1] = c;
		break;
	case 'f':
	case 'e':
	case 'E':
	case 'g':
	case 'G':
		f = floating(nextarg());
		c = fp[1];
		fp[1] = '\0';
		switch (star) {
		case 3:
			printf(fmt, width, prec, f);
			break;
		case 2:
			printf(fmt, prec, f);
			break;
		case 1:
			printf(fmt, width, f);
			break;
		default:
			printf(fmt, f);
		}
		fp[1] = c;
		break;
	case 's':
		c = fp[1];
		fp[1] = '\0';
		sp = nextarg();
		switch (star) {
		case 3:
			printf(fmt, width, prec, sp);
			break;
		case 2:
			printf(fmt, prec, sp);
			break;
		case 1:
			printf(fmt, width, sp);
			break;
		default:
			printf(fmt, sp);
		}
		fp[1] = c;
		break;
	default:
		putchar(*fp & 0377);
		return;
	}
}

int
main(int argc, char **argv)
{
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	progname = basename(argv[0]);
	if (argc > 1 && argv[1][0] == '-' && argv[1][1] == '-' &&
			argv[1][2] == '\0') {
		argv++;
		argc--;
	}
	if (argc <= 1)
		usage();
	ac = argc;
	av = argv;
	a = ab = 2;
	do {
		for (fp = av[1]; *fp; fp++) {
			switch (c = *fp & 0377) {
			case '\\':
				backslash(0, 1);
				break;
			case '%':
				percent();
				break;
			default:
				putchar(c);
			}
		}
	} while (a > ab && a < ac && dolflag == 0);
	if (ferror(stdout))
		status |= 1;
	return status;
}
